### PR TITLE
Implementation of Sparse Coding in Python

### PR DESCRIPTION
**Pull request by natoromano**
This adds a Python implementation of Bruno Olshausen's Sparse Coding algorithm. It is the same as the one described in [this paper](http://redwood.psych.cornell.edu/papers/olshausen_field_1997.pdf), except it uses Locally Competitive Algorithm to compute activations, instead of gradient descent, as described by [Rozell et al.](http://www.ece.rice.edu/~dhj/rozell_icip2007.pdf). The code is inspired by [a lab in Olshausen's Neural Computation course](http://redwood.berkeley.edu/vs265/lab5.pdf) (Matlab).

The file has two classes:
- **SparseNet** is the regular implementation, and can encode regular data types, using batches of data for training. Its public methods are `train`, which is called with the entire data set, `encode` which is called with a single or multiple data points to encode after training, as well as `plotLoss` and `plotBasis` for visualization.

- **ImageSparseNet** (which sub-classes SparseNet) is more in line with the version proposed by Olshausen's lab, specific for natural images. It's particularity is that it uses sub-parts of images of arbitrary size for training (one batch corresponds to sub-parts of one image). It also provides two additional methods, `loadMatlabImages` and `loadNumpyImages`, which allow loading images from .mat files or using NumPy.

Here is a sample output of the produced basis on[ Olshausen's dataset](http://redwood.berkeley.edu/bruno/sparsenet/IMAGES.mat):

@subutai @scottpurdy

### CODE REVIEW

**Comment by subutai** [comment]
@natoromano Following our conventions, can you put each class in its own file?  Also please put a self-contained example script, along with the required data, into projects/sparse_net?  We should be able to replicate the results just by running that script.

**Comment by natoromano** [comment]
@subutai Done.

**Comment by ywcui1990** [comment]
@natoromano Is the new script natural_scenes_sparse_coding.py supposed to reproduce the results of "sparsenet.m" in Bruno's code? I noticed several differences. 

First, A few basis function are unstructured using your script (e.g, 2nd and 3rd basis functions on the 7th row), where as almost all the basis functions are structured using Bruno's code (see below)
image
 
Second, the loss_history.png figure look quite different. Your script gives much lower overall loss (~0.5) whereas Bruno's script gives a loss of around 1.5 after 1000 iterations. 

I will also provide detailed feedback on your code.

**Comment by natoromano** [comment]
@ywcui1990 the loss difference is due to a parameter which is different in Bruno's default code and my sample script (minThreshold, called lambda in Bruno's code, is 0.1 for me and 1.0 in my case). Changing it to 1.0 leads to a loss of ~1.5 after 1000 iterations. However, it still has a couple of unstructured basis. Looking into it.

**Comment by ywcui1990** [comment]
OK, Let's try to use the same parameter values and reproduce their results (at least qualitatively). You might need to increase the number of iterations a bit. I found that using 1000 iterations does not always work with Bruno's code either. They used iteration=10000 in their code.

**Comment by natoromano** [comment]
@ywcui1990  Indeed, training it for 10,000 iterations gets rid of the unstructured basis:



The loss is extremely noisy as expected but averages to about 1.5.

**Comment by natoromano** [comment]
This is the output using 128 basis functions:

File: htmresearch/algorithms/sparse_net.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+               lcaLearningRate=0.1,
+               thresholdDecay=0.95,
+               minThreshold=0.1,
+               thresholdType='soft',
+               verbosity=0,
+               showEvery=500,
+               seed=42):
+    """
+    Initializes the SparseNet.
+    :param inputDim:                (int)   (Flattened) dimension of input data
```
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: Is this dimension of the input data (image)? Or is it the dimension of each filter?

File: htmresearch/algorithms/sparse_net.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Implementation of Bruno Olshausen's sparse coding algorithm.
+It relies on the formulation developed by Olshausen and Field (1996), slightly
+modified so that it uses a Locally Competitive Algorithm (LCA) to compute the
+coefficients, rather than a vanilla gradient descent, as proposed by Rozell
+et al. (2008).
+
```
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: Is it possible to briefly summarize the algorithm here (in addition to the reference)?

File: htmresearch/algorithms/sparse_net.py (hunk 1 above)
Comment (natoromano) [inline htmresearch/algorithms/sparse_net.py]: The name may not be really clear in that regard. I wanted to make an implementation that would not only work for images, and I figured in the general case training batches are not sub-patches but actual batches of data points, each of them being of size inputDim.

Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: We should either rename it or change the description in the doc then. It is not really the dimension of the input data (e.g., see _getDataBatch)

Comment (natoromano) [inline htmresearch/algorithms/sparse_net.py]: OK. Then I guess the _getDataBatch method in the base class should be abstract? And be implemented in sub-classes specific to data types?

File: htmresearch/algorithms/sparse_net.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    if reset:
+      self._reset()
+
+    for _ in xrange(self.numIterations):
+      self._iteration += 1
+
+      batch = self._getDataBatch(inputData)
+
+      # check input dimension, change if necessary
+      if batch.shape[0] != self.inputDim:
```
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: Looks like inputDim can be modified here. Is it really a parameter the user set upfront then?

File: htmresearch/algorithms/image_sparse_net.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    Images are flattened to get a 2-dimensional batch.
+    """
+    if not hasattr(self, 'numImages'):
+      self._initializeDimensions(inputData)
+
+    batch = np.zeros((self.inputDim, self.batchSize))
+    miniImageSize = int(np.sqrt(self.inputDim))
+    # choose random image
+    imageIdx = np.random.choice(range(self.numImages))
```
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: Is it true that the entire data batch is from a single randomly selected image? Why not using more than one images?

File: htmresearch/algorithms/image_sparse_net.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        # pick random starting column
+        cols = self.imageWidth - miniImageSize - 2 * self.BUFFER
+        colNumber = self.BUFFER + np.random.choice(range(cols))
+
+        if self.numChannels is None:
+          miniImage = inputData[rowNumber : rowNumber + miniImageSize,
+                                colNumber : colNumber + miniImageSize,
+                                imageIdx]
+        else:
+          # pick random channel
```
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: I am not sure this is the right way to handle multiple channels. If the image has more than one channel (say RGB), shouldn't the miniImage also be RGB?

File: htmresearch/algorithms/image_sparse_net.py (hunk 4 above)
Comment (natoromano) [inline htmresearch/algorithms/image_sparse_net.py]: Algorithmically, this first pass of the implementation is making the same choices as Bruno. That's what he does in his code. If we believe we should sample from multiple images, that's an easy change. Should I do it?

File: htmresearch/algorithms/sparse_net.py (hunk 1 above)
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: I think that would make more sense. Let's make it abstract in the base class.

File: htmresearch/algorithms/image_sparse_net.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+class ImageSparseNet(SparseNet):
+  """
+  Implementation of SparseNet specifically suited for natural images.
+
+  It's particularity is that a training batch is composed with patches of one
+  particular input image.
+  """
+
+  BUFFER = 4
```
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: What is this BUFFER for? Please document it.

File: htmresearch/algorithms/image_sparse_net.py (hunk 5 above)
Comment (natoromano) [inline htmresearch/algorithms/image_sparse_net.py]: I think both would work (since we are sampling many of those mini images, sampling the channel each time). But I can also take all channels and flatten them. Then filters would also have 3 channels right?

File: htmresearch/algorithms/image_sparse_net.py (hunk 4 above)
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: Hmm, I am not sure why they only sample from one image. Do you see any reason of doing that in their code?

Comment (natoromano) [inline htmresearch/algorithms/image_sparse_net.py]: No, I have never seen this before. And this may be why so many iterations are necessary. Perhaps the idea is to do some stochastic version of convolution over the image?

File: htmresearch/algorithms/sparse_net.py [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+    # update basis
+    gradBasis = batchResiduals.dot(activations.T) / self.batchSize
+    self.basis += self.learningRate * gradBasis
+
+    # normalize basis
+    self.basis /= np.sqrt(np.sum(self.basis ** 2, axis=0))
+
+
+  def _activation(self, input, threshold, thresholdType=None):
```
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: The name of this function is not very descriptive and may be confused with encode (which also gives activation). Maybe to call it thresholdNonlinearity instead?

File: htmresearch/algorithms/sparse_net.py [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    # remove ticks
+    plt.gca().xaxis.set_major_locator(plt.NullLocator())
+    plt.gca().yaxis.set_major_locator(plt.NullLocator())
+    plt.title("Basis functions for {0}".format(self))
+
+    if filename is not None:
+      plt.savefig(filename)
+
+
+  def _reset(self):
```
Comment (ywcui1990) [inline htmresearch/algorithms/sparse_net.py]: In addition to reset, it might be worth adding read/write serialization functionalities. We may need to save the basis and use it later.

File: htmresearch/algorithms/image_sparse_net.py (hunk 5 above)
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: I don't think we need to support color images for now. But if we do, I think the filters either need to contain 3 channels or we have a separate set of filters for each channel. If you pick a random channel here, how would you know which channel the filters are corresponding to when you encode the images? You can't assume the three channels are similar up front.

File: htmresearch/algorithms/image_sparse_net.py (hunk 4 above)
Comment (ywcui1990) [inline htmresearch/algorithms/image_sparse_net.py]: I think it doesn't matter for this homework because they have a small number of large images in the dataset anyway. Imagine you have a large number of small images, you will definitely want to sample from multiple images at each iteration.

**Comment by ywcui1990** [comment]
Done with this round of review. The code is nice and clean in general. @subutai  We verified that we can reproduce Bruno's result with this code.

**Comment by subutai** [comment]
@natoromano @ywcui1990 Great, thanks! Please go ahead and merge.

**Comment by natoromano** [comment]
@ywcui1990 I pushed changes incorporating your comments (serialization/deserialization, filterDim instead of inputDim, algorithm documentation, random images in batch, better handling of multiple channels, some slight refactoring).

**Comment by ywcui1990** [comment]
@natoromano Thanks! Could you also add an example use of the new serialization/deserailzation functionality in the sample code?

**Comment by natoromano** [comment]
@ywcui1990 done.

**Comment by ywcui1990** [comment]
:thumbsup:
